File: solutions/svrp.cpp
// Stochastic Vehicle Routing Problem (SVRP) 
#include <iostream>
#include <fstream>
#include <vector>
#include <cmath>
#include <ctime>
#include <cstdlib>
#include <limits>

using namespace std;

const int NUM_VEHICLES = 3;
const int DEPOT_INDEX = 0;
const double INF = numeric_limits<double>::infinity();

struct Customer {
    int x;
    int y;
};

struct Solution {
    vector<vector<int>> routes;
    double cost;
};

double euclideanDistance(int x1, int y1, int x2, int y2) {
    return sqrt(pow(x1 - x2, 2) + pow(y1 - y2, 2));
}

vector<Customer> readCustomersFromFile(const string& filename) {
    vector<Customer> customers;
    ifstream file(filename);
    if (!file) {
        cerr << "Error: Unable to open file " << filename << endl;
        exit(1);
    }

    int numCustomers;
    file >> numCustomers;

    customers.resize(numCustomers);

    for (int i = 0; i < numCustomers; ++i) {
        file >> customers[i].x >> customers[i].y;
    }

    file.close();
    return customers;
}

Solution generateInitialSolution(const vector<Customer>& customers) {
    Solution initialSolution;
    initialSolution.routes.resize(NUM_VEHICLES);

    for (int i = 1; i < customers.size(); ++i) {
        int vehicle = rand() % NUM_VEHICLES;
        initialSolution.routes[vehicle].push_back(i);
    }

    initialSolution.cost = 0.0;
    for (int v = 0; v < NUM_VEHICLES; ++v) {
        if (!initialSolution.routes[v].empty()) {
            int prevNode = DEPOT_INDEX;
            for (int i = 0; i < initialSolution.routes[v].size(); ++i) {
                int customer = initialSolution.routes[v][i];
                initialSolution.cost += euclideanDistance(customers[prevNode].x, customers[prevNode].y,
                                                          customers[customer].x, customers[customer].y);
                prevNode = customer;
            }
            initialSolution.cost += euclideanDistance(customers[prevNode].x, customers[prevNode].y,
                                                      customers[DEPOT_INDEX].x, customers[DEPOT_INDEX].y);
        }
    }

    return initialSolution;
}

Solution generateNeighborSolution(const Solution& currentSolution) {
    Solution neighborSolution = currentSolution;

    int vehicle1 = rand() % NUM_VEHICLES;
    int vehicle2 = rand() % NUM_VEHICLES;
    while (vehicle1 == vehicle2) {
        vehicle2 = rand() % NUM_VEHICLES;
    }

    if (!neighborSolution.routes[vehicle1].empty() && !neighborSolution.routes[vehicle2].empty()) {
        int customer1 = rand() % neighborSolution.routes[vehicle1].size();
        int customer2 = rand() % neighborSolution.routes[vehicle2].size();

        int temp = neighborSolution.routes[vehicle1][customer1];
        neighborSolution.routes[vehicle1][customer1] = neighborSolution.routes[vehicle2][customer2];
        neighborSolution.routes[vehicle2][customer2] = temp;

        neighborSolution.cost = 0.0;
        for (int v = 0; v < NUM_VEHICLES; ++v) {
            if (!neighborSolution.routes[v].empty()) {
                int prevNode = DEPOT_INDEX;
                for (int i = 0; i < neighborSolution.routes[v].size(); ++i) {
                    int customer = neighborSolution.routes[v][i];
                    neighborSolution.cost += euclideanDistance(customers[prevNode].x, customers[prevNode].y,
                                                               customers[customer].x, customers[customer].y);
                    prevNode = customer;
                }
                neighborSolution.cost += euclideanDistance(customers[prevNode].x, customers[prevNode].y,
                                                           customers[DEPOT_INDEX].x, customers[DEPOT_INDEX].y);
            }
        }
    }

    return neighborSolution;
}

Solution simulatedAnnealing(const vector<Customer>& customers, double initialTemperature, double coolingRate, int iterations) {
    Solution currentSolution = generateInitialSolution(customers);
    Solution bestSolution = currentSolution;

    double temperature = initialTemperature;

    for (int i = 0; i < iterations; ++i) {
        Solution neighborSolution = generateNeighborSolution(currentSolution);
        double deltaCost = neighborSolution.cost - currentSolution.cost;

        if (deltaCost < 0 || exp(-deltaCost / temperature) > ((double) rand() / RAND_MAX)) {
            currentSolution = neighborSolution;
        }

        if (currentSolution.cost < bestSolution.cost) {
            bestSolution = currentSolution;
        }

        temperature *= coolingRate;
    }

    return bestSolution;
}

void outputSolution(const Solution& bestSolution, const vector<Customer>& customers) {
    cout << "Best Solution (Cost = " << bestSolution.cost << "):" << endl;
    for (int v = 0; v < NUM_VEHICLES; ++v) {
        cout << "Vehicle " << v + 1 << ": ";
        for (int i = 0; i < bestSolution.routes[v].size(); ++i) {
            int customerIndex = bestSolution.routes[v][i];
            cout << customerIndex << " ";
        }
        cout << endl;
    }

    for (int v = 0; v < NUM_VEHICLES; ++v) {
        cout << "Route for Vehicle " << v + 1 << ": ";
        if (!bestSolution.routes[v].empty()) {
            int prevNode = DEPOT_INDEX;
            for (int i = 0; i < bestSolution.routes[v].size(); ++i) {
                int customerIndex = bestSolution.routes[v][i];
                cout << prevNode << " -> " << customerIndex << " ";
                prevNode = customerIndex;
            }
            cout << "-> " << DEPOT_INDEX;
        }
        cout << endl;
    }
}

int main() {
    srand(static_cast<unsigned int>(time(nullptr)));

    string filename = "customers.txt";
    vector<Customer> customers = readCustomersFromFile(filename);

    double initialTemperature = 1000;
    double coolingRate = 0.95;
    int iterations = 10000;

    Solution bestSolution = simulatedAnnealing(customers, initialTemperature, coolingRate, iterations);

    outputSolution(bestSolution, customers);

    return 0;
}

File: visualization/svrp.m
%% Stochastic Vehicle Routing Problem (SVRP)
NUM_VEHICLES = 3;
initialTemperature = 1000;
coolingRate = 0.95;
iterations = 10000;
coordinates = [
    10, 20;
    15, 25;
    30, 10;
    25, 15;
    40, 30;
    35, 35;
    20, 5;
    5, 15;
    10, 40
];
x = coordinates(:, 1);
y = coordinates(:, 2);
NUM_CUSTOMERS = size(coordinates, 1);
figure;
hold on;
scatter(x, y, 'filled', 'DisplayName', 'Customers');
scatter(x(1), y(1), 'filled', 'MarkerEdgeColor', 'r', 'MarkerFaceColor', 'r', 'DisplayName', 'Depot');
legend('show');
title('SVRP');
rng('shuffle');  
bestSolution = simulatedAnnealing(initialTemperature, coolingRate, iterations, NUM_CUSTOMERS, NUM_VEHICLES, x, y);
for v = 1:NUM_VEHICLES
    route = bestSolution.routes{v};
    if ~isempty(route)
        route = [1; route(:); 1]; 
        plot(x(route), y(route), '-o', 'LineWidth', 1.5, 'MarkerSize', 8, ...
            'DisplayName', sprintf('Vehicle %d', v));
    end
end
legend('show');
hold off;
function bestSolution = simulatedAnnealing(initialTemperature, coolingRate, iterations, NUM_CUSTOMERS, NUM_VEHICLES, x, y)
    DEPOT_INDEX = 1; 
    bestSolution.routes = cell(NUM_VEHICLES, 1);
    bestSolution.cost = inf;
    currentSolution = generateInitialSolution(NUM_CUSTOMERS, NUM_VEHICLES, x, y);
    temperature = initialTemperature;
    for i = 1:iterations
        neighborSolution = generateNeighborSolution(currentSolution, NUM_VEHICLES, x, y);
        deltaCost = neighborSolution.cost - currentSolution.cost;
        if deltaCost < 0 || exp(-deltaCost / temperature) > rand()
            currentSolution = neighborSolution;
        end
        if currentSolution.cost < bestSolution.cost
            bestSolution = currentSolution;
        end
        temperature = temperature * coolingRate;
    end
end
function initialSolution = generateInitialSolution(NUM_CUSTOMERS, NUM_VEHICLES, x, y)
    DEPOT_INDEX = 1; 
    initialSolution.routes = cell(NUM_VEHICLES, 1);
    initialSolution.cost = 0;
    initialSolution.routes{1} = [2, 5, 7];
    initialSolution.routes{2} = [3, 4];
    initialSolution.routes{3} = [6, 8, 9];
    for v = 1:NUM_VEHICLES
        route = initialSolution.routes{v};
        if ~isempty(route)
            prevNode = DEPOT_INDEX;
            for i = 1:length(route)
                customer = route(i);
                initialSolution.cost = initialSolution.cost + ...
                    euclideanDistance(x(prevNode), y(prevNode), x(customer), y(customer));
                prevNode = customer;
            end
            initialSolution.cost = initialSolution.cost + ...
                euclideanDistance(x(prevNode), y(prevNode), x(DEPOT_INDEX), y(DEPOT_INDEX));
        end
    end
end
function neighborSolution = generateNeighborSolution(currentSolution, NUM_VEHICLES, x, y)
    DEPOT_INDEX = 1; 
    neighborSolution = currentSolution;
    vehicle1 = randi(NUM_VEHICLES);
    vehicle2 = randi(NUM_VEHICLES);
    while vehicle1 == vehicle2
        vehicle2 = randi(NUM_VEHICLES);
    end
    route1 = neighborSolution.routes{vehicle1};
    route2 = neighborSolution.routes{vehicle2};
    if ~isempty(route1) && ~isempty(route2)
        customer1 = randi(length(route1));
        customer2 = randi(length(route2));
        temp = route1(customer1);
        route1(customer1) = route2(customer2);
        route2(customer2) = temp;
        neighborSolution.routes{vehicle1} = route1;
        neighborSolution.routes{vehicle2} = route2;
        neighborSolution.cost = 0;
        for v = 1:NUM_VEHICLES
            route = neighborSolution.routes{v};
            if ~isempty(route)
                prevNode = DEPOT_INDEX;
                for i = 1:length(route)
                    customer = route(i);
                    neighborSolution.cost = neighborSolution.cost + ...
                        euclideanDistance(x(prevNode), y(prevNode), x(customer), y(customer));
                    prevNode = customer;
                end
                neighborSolution.cost = neighborSolution.cost + ...
                    euclideanDistance(x(prevNode), y(prevNode), x(DEPOT_INDEX), y(DEPOT_INDEX));
            end
        end
    end
end
function distance = euclideanDistance(x1, y1, x2, y2)
    distance = sqrt((x1 - x2)^2 + (y1 - y2)^2);
end
